File: java-devildyw/src/main/java/top/dyw/leetcode/面试150题/L206反转链表.java
package top.dyw.leetcode.面试150题;

/**
 * @author Devildyw
 * @date 2023/08/23 22:22
 **/
@SuppressWarnings("all")
public class L206反转链表 {

    public static void main(String[] args) {

    }

    public ListNode reverseList(ListNode head) {
        ListNode pre = null;
        ListNode cur = head;
        while (cur!=null){
            ListNode temp = cur.next;
            cur.next = pre;
            pre = cur;
            cur = temp.next;
        }

        return pre;
    }

    class ListNode {
        int val;
        ListNode next;

        ListNode() {
        }

        ListNode(int val) {
            this.val = val;
        }

        ListNode(int val, ListNode next) {
            this.val = val;
            this.next = next;
        }
    }
}

File: golang-devildyw/simple/L206反转链表.go
package main

func main() {

}

func reverseList(head *ListNode) *ListNode {
	var (
		pre *ListNode = nil
		p             = head
	)

	for p != nil {
		next := p.Next
		p.Next = pre
		pre = p
		p = next
	}
	return pre
}

func reverseList_TwoPointer(head *ListNode) *ListNode {
	var (
		cur *ListNode = nil
		pre           = head
	)

	for pre != nil {
		next := pre.Next
		pre.Next = cur
		cur = pre
		pre = next
	}
	return cur
}
